source/solvers/navier_stokes_assemblers.cc

```diff
@@ -0,0 +1,675 @@
+    {
+      // Velocity
+      const Tensor<1, dim> velocity    = scratch_data.velocity_values[q];
+      const double velocity_divergence = scratch_data.velocity_divergences[q];
+      const Tensor<2, dim> velocity_gradient =
+        scratch_data.velocity_gradients[q];
+      const Tensor<1, dim> velocity_laplacian =
+        scratch_data.velocity_laplacians[q];
+
+      // Pressure
+      const double         pressure = scratch_data.pressure_values[q];
+      const Tensor<1, dim> pressure_gradient =
+        scratch_data.pressure_gradients[q];
+
+      // Forcing term
+      const Tensor<1, dim> force = scratch_data.force[q];
+
+      // Calculation of the magnitude of the
+      // velocity for the stabilization parameter
+      const double u_mag = std::max(velocity.norm(), 1e-12);
+
+      // Store JxW in local variable for faster access;
+      const double JxW = JxW_vec[q];
+
+      // Calculation of the GLS stabilization parameter. The
+      // stabilization parameter used is different if the simulation
+      // is steady or unsteady. In the unsteady case it includes the
+      // value of the time-step
+      const double tau =
+        this->simulation_control->get_assembly_method() ==
+            Parameters::SimulationControl::TimeSteppingMethod::steady ?
+          1. / std::sqrt(std::pow(2. * u_mag / h, 2) +
+                         9 * std::pow(4 * viscosity / (h * h), 2)) :
+          1. / std::sqrt(std::pow(sdt, 2) + std::pow(2. * u_mag / h, 2) +
+                         9 * std::pow(4 * viscosity / (h * h), 2));
+
+
+      // Calculate the strong residual for GLS stabilization
+      auto strong_residual = velocity_gradient * velocity + pressure_gradient -
+                             viscosity * velocity_laplacian - force +
+                             strong_residual_vec[q];
+
+      // Assembly of the right-hand side
+      for (unsigned int i = 0; i < n_dofs; ++i)
+        {
+          const auto phi_u_i      = scratch_data.phi_u[q][i];
+          const auto grad_phi_u_i = scratch_data.grad_phi_u[q][i];
+          const auto phi_p_i      = scratch_data.phi_p[q][i];
+          const auto grad_phi_p_i = scratch_data.grad_phi_p[q][i];
+          const auto div_phi_u_i  = scratch_data.div_phi_u[q][i];
+
+
+          // Navier-Stokes Residual
+          local_rhs(i) +=
+            (
+              // Momentum
+              -viscosity * scalar_product(velocity_gradient, grad_phi_u_i) -
+              velocity_gradient * velocity * phi_u_i + pressure * div_phi_u_i +
+              force * phi_u_i -
+              // Continuity
+              velocity_divergence * phi_p_i) *
+            JxW;
+
+          // PSPG GLS term
+          local_rhs(i) += -tau * (strong_residual * grad_phi_p_i) * JxW;
+
+          // SUPG GLS term
+          if (SUPG)
+            {
+              local_rhs(i) +=
+                -tau * (strong_residual * (grad_phi_u_i * velocity)) * JxW;
+            }
```

Comment:
Could we use the same strategy as with the matrix i.e. writing to a double variable and the writing to the `local_rhs(i)` once? Or it does not speed up the computation at all in this case?

Reply:
You are right we could. It will most likely have a more minor effect, but worth trying! I'll add this.

Reply:
Added :)!